test/unit/org/apache/cassandra/tools/nodetool/CompactionHistoryTest.java

```diff
@@ -0,0 +1,133 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.tools.nodetool;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Lists;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import org.apache.cassandra.cql3.CQLTester;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.db.compaction.OperationType;
+import org.apache.cassandra.tools.ToolRunner;
+import org.apache.cassandra.utils.FBUtilities;
+
+import static org.apache.cassandra.db.compaction.CompactionHistoryTabularData.COMPACTION_TYPE_PROPERTY;
+import static org.apache.cassandra.tools.ToolRunner.invokeNodetool;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(Parameterized.class)
+public class CompactionHistoryTest extends CQLTester
+{
+    @Parameterized.Parameter
+    public List<String> cmd;
+
+    @Parameterized.Parameter(1)
+    public String compactionType;
+    
+    @Parameterized.Parameter(2)
+    public int systemTableRecord;
+
+    @Parameterized.Parameters(name = "{index}: cmd={0} compactionType={1} systemTableRecord={2}")
+    public static Collection<Object[]> data()
+    {
+        List<Object[]> result = new ArrayList<>();
+        result.add(new Object[]{ Lists.newArrayList("compact"), OperationType.MAJOR_COMPACTION.type, 1 });
+        result.add(new Object[]{ Lists.newArrayList("garbagecollect"), OperationType.GARBAGE_COLLECT.type, 20 });
+        result.add(new Object[]{ Lists.newArrayList("upgradesstables", "-a" ), OperationType.UPGRADE_SSTABLES.type, 10 });
+        return result;
+    }
+    
+    @BeforeClass
+    public static void setup() throws Exception
+    {
+        requireNetwork();
+        startJMXServer();
+    }
+    
+    @Test
+    public void testCompactionProperties() throws Throwable
+    {
+        createTable("CREATE TABLE %s (id text, value text, PRIMARY KEY ((id)))");
+        ColumnFamilyStore cfs = Keyspace.open(keyspace()).getColumnFamilyStore(currentTable());
+        cfs.disableAutoCompaction();
+        // write SSTables for the specific key
+        for (int i = 0; i < 10; i++)
+        {
+            execute("INSERT INTO %s (id, value) VALUES (?, ?)", "key" + i, "value" + i);
+            flush(keyspace());
+            if(compactionType.equals(OperationType.GARBAGE_COLLECT.type))
```

Comment:
just curious, aren't deletions valid for other types of compaction?

Reply:
@jacek-lewandowski yes you are right, I think that `garbagecollect` is used after cluster expansion and it is executed when data are not held by that respective node anymore as token assignment has changed. However, for the sake of testing this, it is ok.

What I do is that I will remove this logic and we will not need the below code either.